refactor: switch from optparse to argparse

vergenzt opened the pull request:
To consolidate options definitions into a single place, and use the more modern Python option parsing library.

nedbat commented on cogapp/test_cogapp.py:
```diff
@@ -1046,7 +1053,7 @@ def test_print_output(self):
         self.cog.callable_main(["argv0", "-rP", "test.cog"])
         self.assertFilesSame("test.cog", "test.out")
-        output = self.output.getvalue()
+        output, errput = self.capsys.readouterr()
```
throughout: if we aren't using `errput`, we should assign `output, _ = ...`, or we should make assertions about errput.

nedbat commented on cogapp/options.py:
```diff
@@ -0,0 +1,307 @@
+    _parser: ClassVar = CogArgumentParser(
+        prog="cog",
+        usage=argparse.SUPPRESS,
+        description=rewrap(description),
```
What happens if we don't use rewrap?

LecrisUT replied:
In my experience `click` handles this stuff quite automagically, but it's an issue if you would want such a dependency or not. The desiyusing that is quite different as well, but IMO it is cleaner and more readable.

vergenzt replied:
Removed the rewrapping

vergenzt commented on cogapp/__init__.py:
```diff
@@ -5,3 +5,7 @@
 from .cogapp import Cog as Cog, CogUsageError as CogUsageError, main as main
+
+
+if __name__ == "__main__":
```
This was here to enable calling the app as `python -m cogapp <cli_args>`. https://docs.python.org/3/library/__main__.html#idiomatic-usage

I used it while testing out the `--help` output and figured it would be useful to keep. I'm fine taking it out if you don't want it there though.

LecrisUT replied:
There is already a `__main__.py` that handles that

vergenzt replied:
Removed

vergenzt commented on cogapp/options.py:
```diff
@@ -0,0 +1,307 @@
+
+
+if sys.version_info >= (3, 8):
+    from argparse import _ExtendAction
```
The supported way to do it is by passing the string "extend", but that's only available starting in Python 3.8. I'll refactor to use the string in that case though.

vergenzt commented on cogapp/options.py:
```diff
@@ -0,0 +1,307 @@
+            width=HELP_WIDTH,
+        )
+
+    def _split_lines(self, text, width):
```
We don't have to necessarily. I wasn't thinking of it as problematic but you make a good point. It was purely for aesthetic reasons to add the blank line between [option descriptions](https://github.com/nedbat/cog/pull/38/files#diff-b26d76f06d0960fbd1b40f06e391995887e1e3e12d5b4b2b55a516d87ab869c3R40-R88) (which becomes more aesthetically pleasing once there are long options, e.g. see https://github.com/vergenzt/cog/blob/bd3cb2435419906c90ddd08febe8a7df7544260a/docs/running.rst).

I'll remove it though - it's probably not worth the hassle.

vergenzt replied:
Removed

vergenzt commented:
Sorry gtg for now - will reply to remainder ASAP

LecrisUT commented on cogapp/options.py:
```diff
@@ -0,0 +1,307 @@
+        getattr(ns, self.dest).update([arg])
+
+
+class Markers(NamedTuple):
```
👍 for the dataclass suggestion. Looking at how it is used, I don't see that it would be incompatible

vergenzt commented:
FYI I rebased and massively simplified this PR, removing most of the cruft and I believe addressing all comments. There's one failing test in CI which I'm confused about since all tests were passing locally; I'll update that as soon as I can.
